Closes #5565: Helm reconcile period from CR annotations or watches.yaml (#5585)
operator-framework/operator-sdk: website/content/en/docs/building-operators/helm/reference/advanced_features/reconcile-period.md

---
title: Reconcile Period from Custom Resource Annotations
linkTitle: Custom Resource Annotations Reconcile Period
weight: 200
description: Allow a user to set the desired reconcile period from the custom resource's annotations
---

> While running a helm operator, we can specify the reconcile-period through the custom resource's annotations under the `helm.operator-sdk/reconcile-period` key. Using this feature will reconcile the cluster in a fixed interval of time. The reconcile period can be specified in the custom resource's annotations in the following manner: 

Reviewer: ```suggestion
While running a Helm-based operator, the reconcile-period can be specified through the custom resource's annotations under the `helm.operator-sdk/reconcile-period` key. Using this feature will reconcile the cluster in a fixed interval of time. The reconcile period can be specified in the custom resource's annotations in the following manner: 
```

> While running a helm operator, we can specify the reconcile-period through the custom resource's annotations under the `helm.operator-sdk/reconcile-period` key. Using this feature will reconcile the cluster in a fixed interval of time. The reconcile period can be specified in the custom resource's annotations in the following manner: 

Reviewer: Also, does having a reconcile-period actually mean it has a fixed interval reconcile? I thought the reconcile periods were a MINIMUM meaning if no reconcile has happened then run a reconcile. If events come in before that period will the operator not reconcile?

Author: Thanks for calling that out. You're right - this ensures that at a minimum the cluster will reconcile in the specified interval of time. However, if events come in between before the period elapses, the cluster **will** get reconciled (as it should). This feature is to "ensure" that the cluster does not go long periods of time without reconciling. 

I've updated the description in the docs with that explanation.

```sh
...
metadata:
  name: nginx-sample
  annotations:

>     helm.operator-sdk/reconcile-period: "5s"

Reviewer 2: What happens if the reconciliation period not be informed?
What is the default behaviour? 
Have we the same behaviour for Ansible and Helm in this case?

Author: If no reconcile period is specified in the CR annotations, then the default value is 1m as indicated [here](https://github.com/operator-framework/operator-sdk/blob/master/internal/helm/flags/flag.go#L60).

Yes, this behavior exists for ansible.

Reviewer 2: Can be a follow up either, but should we not. ensure that same for Ansible?

Reviewer 3: @camilamacedo86: So we are now going to accept reconcile Period through 3 ways:
1. from CR annotations 
3. command line flags 
3. through watches.yaml 
The order of precedence imo should be CR annotations > watches.yaml > cmd line flags. We want it to be in sync with ansible, but ansible does not accept it from flags and neither does the hybrid helm implementation. We would want to eventually remove accepting it through cmd line flags, but I think that should be a follow up. From discussion with @VenkatRamaraju, looks like CR annotations override values from `watches.yaml` in ansible. So I think we are good with that.

...
```

The value that is present under this key must be in the h/m/s format. For example, 1h2m4s, 3m0s, 4s are all valid values, but 1x3m9s is invalid. 

> **NOTE**: This is just one way of specifying the reconcile period for helm operators. Another way is using the `--reconcile-period` command line flag while running the operator. However, if both

Reviewer: ```suggestion
**NOTE**: This is just one way of specifying the reconcile period for Helm-based operators. Another way is using the `--reconcile-period` command line flag while running the operator. However, if both
```

> of these methods are used together (which they shouldn't be), the custom resource's annotations will take precedence. 

Reviewer: Need to make sure we call attention to it.
```suggestion
of these methods are used together (which they shouldn't be), the *custom resource's annotations will take precedence*. 
```
